go/libraries/doltcore/sqle/dsess/dolt_session.go

```diff
@@ -0,0 +1,211 @@
+func validatePersistableSysVar(name string) (sql.SystemVariable, interface{}, error) {
+	sysVar, val, ok := sql.SystemVariables.GetGlobal(name)
+	if !ok {
+		return sql.SystemVariable{}, nil, sql.ErrUnknownSystemVariable.New(name)
+	}
+	if !sysVar.Dynamic {
+		return sql.SystemVariable{}, nil, sql.ErrSystemVariableReadOnly.New(name)
+	}
+	return sysVar, val, nil
+}
+
+// getPersistedValue reads and converts a config value to the associated SystemVariable type
+func getPersistedValue(conf config.ReadableConfig, k string) (interface{}, error) {
+	v, err := conf.GetString(k)
+	if err != nil {
+		return nil, err
+	}
+
+	_, value, err := validatePersistableSysVar(k)
+	if err != nil {
+		return nil, err
+	}
+
+	var res interface{}
+	switch value.(type) {
+	case int, int8, int16, int32, int64:
+		res, err = strconv.ParseInt(v, 10, 64)
+	case uint, uint8, uint16, uint32, uint64:
+		res, err = strconv.ParseUint(v, 10, 64)
+	case float32, float64:
+		res, err = strconv.ParseFloat(v, 64)
+	case string:
+		return v, nil
+	default:
+		return nil, sql.ErrInvalidType.New(value)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// setPersistedValue casts and persists a key value pair assuming thread safety
+func setPersistedValue(conf config.WritableConfig, key string, value interface{}) error {
+	switch v := value.(type) {
+	case int:
+		return config.SetInt(conf, key, int64(v))
+	case int8:
+		return config.SetInt(conf, key, int64(v))
+	case int16:
+		return config.SetInt(conf, key, int64(v))
+	case int32:
+		return config.SetInt(conf, key, int64(v))
+	case int64:
+		return config.SetInt(conf, key, v)
+	case uint:
+		return config.SetUint(conf, key, uint64(v))
+	case uint8:
+		return config.SetUint(conf, key, uint64(v))
+	case uint16:
+		return config.SetUint(conf, key, uint64(v))
+	case uint32:
+		return config.SetUint(conf, key, uint64(v))
+	case uint64:
+		return config.SetUint(conf, key, v)
+	case float32:
+		return config.SetFloat(conf, key, float64(v))
+	case float64:
+		return config.SetFloat(conf, key, v)
+	case string:
+		return config.SetString(conf, key, v)
+	default:
+		return sql.ErrInvalidType.New(v)
+	}
+}
+
+// NewPersistedSystemVariables returns system variables from the persisted config
+func NewPersistedSystemVariables(conf config.ReadableConfig) ([]sql.SystemVariable, error) {
+	allVars := make([]sql.SystemVariable, conf.Size())
+	i := 0
+	var err error
+	var sysVar sql.SystemVariable
+	var def interface{}
+	conf.Iter(func(k, v string) bool {
+		def, err = getPersistedValue(conf, k)
+		if err != nil {
+			err = fmt.Errorf("key: '%s'; %w", k, err)
+			return true
+		}
+		// getPeristedVal already checked for errors
+		sysVar, _, _ = sql.SystemVariables.GetGlobal(k)
+		sysVar.Default = def
+		allVars[i] = sysVar
+		i++
+		return false
+	})
+	if err != nil {
+		return nil, err
+	}
+	return allVars, nil
+}
+
+func InitPersistedSystemVars(dEnv *env.DoltEnv) error {
```

Comment:
This method makes me uneasy. Seems like calling this more than once, or after an engine is already server queries, is really bad news?

Needs a lock to make sure it's only called once

Reply:
yeah there were some ugly race's on the first pass. adding a global `initMu` lock, lmk if that's too much of an antipattern